commander/tracker/sender.go

```diff
@@ -0,0 +1,88 @@
+package tracker
+
+import (
+	"github.com/Worldcoin/hubble-commander/eth"
+	"github.com/Worldcoin/hubble-commander/models"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/pkg/errors"
+)
+
+var ErrIncorrectTxParamsType = errors.New("incorrect tx params type")
+
+type TxsSender struct {
+	client       *eth.Client
+	requestsChan chan *txRequest
+}
+
+func newTxRequestsSender(ethClient *eth.Client) *TxsSender {
+	return &TxsSender{
+		client:       ethClient,
+		requestsChan: make(chan *txRequest, 32),
+	}
+}
+
+func (t *TxsSender) SubmitTransfersBatch(
+	batchID *models.Uint256,
+	commitments []models.CommitmentWithTxs,
+) (*types.Transaction, error) {
+	return t.sendTxRequest(&submitTransfersBatchRequestParams{
+		batchID:     batchID,
+		commitments: commitments,
+	})
+}
+
+func (t *TxsSender) SubmitCreate2TransfersBatch(
+	batchID *models.Uint256,
+	commitments []models.CommitmentWithTxs,
+) (*types.Transaction, error) {
+	return t.sendTxRequest(&submitCreate2TransfersBatchRequestParams{
+		batchID:     batchID,
+		commitments: commitments,
+	})
+}
+
+func (t *TxsSender) SubmitMassMigrationsBatch(
+	batchID *models.Uint256,
+	commitments []models.CommitmentWithTxs,
+) (*types.Transaction, error) {
+	return t.sendTxRequest(&submitMassMigrationsBatchRequestParams{
+		batchID:     batchID,
+		commitments: commitments,
+	})
+}
+
+func (t *TxsSender) WithdrawStakeRequest(batchID *models.Uint256) (*types.Transaction, error) {
+	return t.sendTxRequest(&withdrawStakeRequestParams{
+		batchID: batchID,
+	})
+}
+
+func (t *TxsSender) sendRequest(request *txRequest) (*types.Transaction, error) {
+	tx, err := t.sendTransaction(request.params)
+	request.responseChannel <- txRequestResponse{Tx: tx, Err: err}
+	return tx, err
+}
+
+func (t *TxsSender) sendTransaction(params interface{}) (*types.Transaction, error) {
+	switch typedParams := params.(type) {
+	case *submitTransfersBatchRequestParams:
```

Comment:
I'm wondering if it wouldn't be easier to encode transactions much earlier - before they are pushed through the channel. This way you'd only have a single generic transaction type with data, value, and recipients address and you wouldn't need to keep repetitive boilerplate for every transaction that we send.

Try seeing if there's a good way to utilise smart-contract bindings to just encode transactions without sending them, and if not, you can try using the raw ABIs for that.